### ISSUE TITLE
Memory release after Parallel

### ISSUE THREAD

**Issue by yangqch** [joblib/joblib#167]
I did a test and found objects returned by parallel processing cannot be del and thus memory cannot be released.
simplified code:

```
### Measure memory usage
import psutil, gc
ps = psutil.Process()
def memory_usage_psutil(ps):
    gc.collect()
    return ps.get_memory_info().rss / (2 ** 20)

### a function making a list from 0 to 10**6
def make_list(i): return [x for x in xrange(10 ** 6)]

print 'start', memory_usage_psutil()
### create 5 such lists w/o parallel and delete
single = [make_list(i) for i in range(5)]
print 'have single list', memory_usage_psutil()
del single
print 'delete single list', memory_usage_psutil()

### create 5 such lists with parallel and delete
parallel = Parallel(n_jobs=5)(delayed(make_list)(i) for i in range(5))
print 'have parallel list', memory_usage_psutil()
del parallel
print 'delete parallel list', memory_usage_psutil()

### =========Output==========
start 19.1796875
have single list 174.328125
delete single list 19.3203125
have parallel list 194.50390625
delete parallel list 154.9609375
```

**Comment by Tych0n** [comment]
Past four years stuck with the same issue - after running joblib's Parallel, deleting results and gc.collect() -ing I still have increased memory (checking by htop for process line). Found no way to free memory back. To broaden @yangqch answer I use such code to isolate memory for parallel computations:

<details><summary>Imports</summary>
<p>

```python
import multiprocessing as mp
n_cores = mp.cpu_count()
import time
import datetime
from sklearn.externals.joblib import Parallel, delayed
import sklearn
from functools import partial
import pickle
```
</p>
</details>
<details><summary>joblib progress bar</summary>
<p>

```python
class BatchCompletionCallBack(object):
    # Added code - start
    global total_n_jobs
    global jobs_start_time
    # Added code - end
    def __init__(self, dispatch_timestamp, batch_size, parallel):
        self.dispatch_timestamp = dispatch_timestamp
        self.batch_size = batch_size
        self.parallel = parallel

    def __call__(self, out):
        self.parallel.n_completed_tasks += self.batch_size
        this_batch_duration = time.time() - self.dispatch_timestamp

        self.parallel._backend.batch_completed(self.batch_size,
                                           this_batch_duration)
        self.parallel.print_progress()
        # Added code - start
        progress = self.parallel.n_completed_tasks / total_n_jobs
        elapsed = int((time.time() - jobs_start_time) / self.parallel.n_completed_tasks * (total_n_jobs - self.parallel.n_completed_tasks))
        print(
            "\r[{:50s}] {:.1f}% - Elapsed time {} ".format(
                '#' * int(progress * 50)
                , progress*100
                , datetime.timedelta(seconds=elapsed))
            , end="", flush=True)
        if self.parallel.n_completed_tasks == total_n_jobs:
            print('\n')
        # Added code - end
        if self.parallel._original_iterator is not None:
            self.parallel.dispatch_next()

sklearn.externals.joblib.parallel.BatchCompletionCallBack = BatchCompletionCallBack
```
</p>
</details>
<details><summary>Parallel wrapper</summary>
<p>

```python
def parallel_wrapper(func, *args, **kwargs):
    global total_n_jobs, jobs_start_time
    total_n_jobs = len(args[0])
    jobs_start_time = time.time()

    if kwargs:
        mapfunc = partial(func, **kwargs)
    else:
        mapfunc = func

    with open('file.pkl', 'wb') as file:
        pickle.dump(Parallel(n_jobs=n_cores)(map(delayed(mapfunc), *args)), file)
    print('Operating time - ', datetime.timedelta(seconds=int(time.time() - jobs_start_time)))
```
</p>
</details>
<details><summary>Class to use in 'with' construction</summary>
<p>

```python
class isolate:
    def __init__(self, func):
        self.func = func

    def __enter__(self):
        return self

    def run(self, func, *args, **kwargs):
        self.p = mp.Process(target=self.func, args=(func, *args), kwargs=kwargs)
        self.p.start()

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.p.join()
```
</p>
</details>
<details><summary>Example</summary>
<p>

```python
# Dummy function we want to parallelize
def dummy(i, keyword_argument=None):
    return (i.shape, keyword_argument)
# Dummy data for example
import numpy as np
tmp = np.ones((5,5))
# Example
with isolate(parallel_wrapper) as isolated:
    isolated.run(dummy, [tmp, tmp], keyword_argument=2)

with open('file.pkl', 'rb') as file:
    tmp = pickle.load(file)

tmp
# Cleaning joblib to remove progress bar
import importlib
importlib.reload(sklearn.externals.joblib.parallel)
```
</p>
</details>

**Comment by tomMoral** [comment]
This issue should be resolved in the latest version of `joblib` which feature memory leaks safeguards with the `loky` backend:

 - If `psutil` is installed on the system, a worker process is shutdown and a new worker is re-spawn if its memory usage grows by more than 100Mb between two tasks (the checks are performed at most once every second).
 - else, we call `gc.collect` periodically between 2 tasks.

Please try to use `joblib==0.12.4` and report if you still see some memory leaked.

**Comment by ogrisel** [comment]
I cannot reproduce the original problem with recent python (3.7) and joblib (0.12.4):

```python
from joblib import Parallel, delayed
import psutil, gc


def memory_usage_psutil():
    gc.collect()
    return psutil.Process().memory_info().rss / (2 ** 20)

### a function making a list from 0 to 10**6
def make_list(i):
    return list(range(10 ** 6))

print('start', memory_usage_psutil())
### create 5 such lists w/o parallel and delete
single = [make_list(i) for i in range(5)]
print('have single list', memory_usage_psutil())
del single
print('delete single list', memory_usage_psutil())

### create 5 such lists with parallel and delete
parallel = Parallel(n_jobs=5, backend='loky')(
    delayed(make_list)(i) for i in range(5))
print('have parallel list (loky)', memory_usage_psutil())
del parallel
print('delete parallel list (loky)', memory_usage_psutil())


### create 5 such lists with parallel and delete
parallel = Parallel(n_jobs=5, backend='multiprocessing')(
    delayed(make_list)(i) for i in range(5))
print('have parallel list (multiprocessing)', memory_usage_psutil())
del parallel
print('delete parallel list (multiprocessing)', memory_usage_psutil())
```

Here is the output:

```
start 30.07421875
have single list 223.16796875
delete single list 30.19921875
have parallel list (loky) 234.33203125
delete parallel list (loky) 30.83984375
have parallel list (multiprocessing) 234.87890625
delete parallel list (multiprocessing) 31.47265625
```

As you can see it's unrelated to loky or multiprocessing. There might have been a bug in joblib in the past but it was probably fixed before 0.12 and the use of loky anyway.

Closing.

**Comment by ogrisel** [comment]
Here is the output with joblib==0.11:

```
start 29.16015625
have single list 222.26171875
delete single list 29.6171875
have parallel list (multiprocessing) 234.0078125
delete parallel list (multiprocessing) 30.515625
```

(no problem either).

**Comment by ogrisel** [comment]
I went back to old versions of joblib 0.8.4. And I cannot reproduce the bug either. It might have been fixed on the Python side (I am using Python 3.7) and is therefore unrelated to joblib as Gael initially suggested..
